### ISSUE TITLE
Password in Browser before you can Open the Garage

### ISSUE THREAD

**Issue by ZyginRNS** [nebhead/garage-zero#24]
Hi i wanted to ask, if you Can protect the Garage door being  Opend With a password or anything like That, because i wanted to make a Little touchpad Outside the Garage, if my mobile phone is Empty Or anything like That

**Comment by nebhead** [comment]
Oh, that's a very interesting idea.  I'm thinking maybe a separate keypad, touchscreen or something, maybe a companion application that can be installed on another device that is mounted outside.  What kind of touch pad were you thinking?  Would you be opposed to using an Arduino device?  

I'm away from my workshop for a few more weeks, but this sounds like something I might want to implement myself.

**Comment by ZyginRNS** [comment]
i thaught about an  pc like an little raspberry pi, with just a little screen and an pinpad where you can only acces the website to open the door

**Comment by nebhead** [comment]
Right, so what I was thinking about was perhaps using an Arduino based system based on an ESP8266 platform so it can communicate over Wi-Fi with the garage door controller, and then run on batteries only when in use.  What we could implement is an API Key system that will allow devices to trigger the garage door button with a specific URL + API Key that can be securely generated.  

So, maybe a D1-Mini (they are really cheap on Amazon) hooked up to a keypad matrix and a couple of LED's. Then have a switch that turns on the system, and waits for keypad input, for a specific PIN code, which triggers an API call.  This can all be hardcoded to the device firmware and would be reasonably secure.  Since the D1-Mini is very low power, we could probably run this from a few AA batteries, and if it's off until a switch is activated (i.e. like the keypad cover being lifted), then it won't consume any power just sitting there.  

It might take a second for the system to boot, but it's better than a raspberry pi or something, which can take over a minute to boot.  Also, it's pretty secure because someone can't just steal it, or hack into it to get it to work.  They'd still have to guess your PIN code to trigger the door open command.  At least I don't think there is a way for someone to dump the firmware on these things.  But honestly, if they went to that much trouble, there are easier ways to get into a garage. :)  

A keypad like this (https://www.amazon.com/dp/B000TA8PE4/) and D1 Mini (https://www.amazon.com/AITRIP-NodeMcu-Internet-Development-Compatible/dp/B08C7FYM5T) some LEDs, a tack switch and a 3D printed enclosure would make this look pretty slick.  Could even put a solar charger on this thing to keep rechargeable batteries topped off.  This is kind getting me excited. :)

**Comment by J0eK1ng** [comment]
I'm all for security and all but if your going to implement a password protected website can you put it a yes/no variable in your code? This would allow us to decide if we want that feature or not.  I don't need or want to have to input a pwd every time i want to open/close my garage door. 
Now I understand this is important for an external keypad but not for the main web site. Just a thought.

**Comment by nebhead** [comment]
Yep, definitely could implement an option to leave this open for those who want to.  It will take me some time, but I will definitely work on this idea.  It looks like in Flask, the session can put a cookie on the device and thus would only need to log in once.  

I've ordered some keypads from Amazon to start playing around with this.
